include/internal/json_enc.h

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef JSON_ENC_H
+# define JSON_ENC_H
+
+# include <openssl/bio.h>
+
+/*
+ * JSON Encoder
+ * ============
+ *
+ * This JSON encoder is used for QLOG. It supports ordinary JSON (RFC 7159),
+ * JSON-SEQ (RFC 7464) and I-JSON (RFC 7493). It supports only basic ASCII.
+ */
+
+struct json_write_buf {
+    BIO     *bio;
+    char    *buf;
+    size_t  alloc, cur;
+};
+
+typedef struct json_enc_st {
+    uint32_t                flags;
+    /* error: 1 if an error has occurred. */
+    /* state: current state. */
+    /* stack stores a bitmap. 0=object, 1=array. */
+    /* stack cur   size: stack_end_byte bytes, stack_end_bit bits. */
+    /* stack alloc size: stack_bytes bytes. */
+    unsigned char           error, stack_end_bit, state, *stack, defer_indent;
+    unsigned char           stack_small[16];
+    struct json_write_buf   wbuf;
+    size_t                  stack_end_byte, stack_bytes;
+} JSON_ENC;
```

Comment:
I would still OSSL_ prefix these - I can see these ending up as public APIs and you named the functions with a prefix but not the typedefs and macros - make them all have a prefix.

There are so many JSON libraries out there so at least not clashing with them is prudent - and it makes promoting things later easier.

Reply:
Fair enough &mdash; the reason for the naming is that symbols do actually have to have the `ossl_` prefix (e.g. to avoid conflicts in statically linked libraries) whereas the same isn't the case for typedefs.

I'll make the change.